src/gwt/src/org/rstudio/core/client/KeyboardTracker.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * KeyboardTracker.java
+ *
+ * Copyright (C) 2009-17 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.core.client;
+
+import org.rstudio.core.client.command.KeyboardShortcut;
+
+import com.google.gwt.dom.client.NativeEvent;
+import com.google.gwt.user.client.Event;
+import com.google.gwt.user.client.Event.NativePreviewEvent;
+import com.google.gwt.user.client.Event.NativePreviewHandler;
+import com.google.inject.Singleton;
+
+@Singleton
+public class KeyboardTracker
+{
+   public KeyboardTracker()
+   {
+      Event.addNativePreviewHandler(new NativePreviewHandler()
+      {
+         @Override
+         public void onPreviewNativeEvent(NativePreviewEvent preview)
+         {
+            int type = preview.getTypeInt();
+            if ((type & Event.KEYEVENTS) == 0)
+               return;
+            
+            NativeEvent event = preview.getNativeEvent();
+            modifier_ = KeyboardShortcut.getModifierValue(event);
+         }
+            
+      });
+   }
+   
+   public boolean isShiftKeyDown()
+   {
+      return (modifier_ & KeyboardShortcut.SHIFT) != 0;
```

Comment:
How does this work with mouse-initiated pastes? Do we get a key event for Shift by itself?

Reply:
We do, fortunately!

There is one wrinkle. If you press and hold Shift _after_ using the right mouse button to open the context menu, then this won't work. This is because the browser 'loses' focus (it's transferred to the context menu) and so the user's Shift keypress isn't registered.

I think we can live with this deficiency, though.